### PR TITLE
fix(http): correctly handle response body for 204 status code

### LINKED ISSUE #12393: Exception is thrown for 204 HTTP response
**Current behavior**
Exception is thrown for 204 HTTP response in case request contains responseType: ResponseContentType.Json

Error: Failed to read the 'responseText' property from 'XMLHttpRequest': The value is only accessible if the object's 'responseType' is '' or 'text' (was 'json').

The exception is thrown from XHRConnection, see: 
var body = isPresent(_xhr.response) ? _xhr.response : _xhr.responseText; in xhr_backend.ts

In this case, _xhr.response is null, and _xhr.responseText does not even exist for 204.

**Expected behavior**
Response should not be attempted to be parsed in case of 204 despite invalid ResponseContentType.
- **Angular version:** 2.0.0
- **Language:** [TypeScript 2.0.2]

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Fixes https://github.com/angular/angular/issues/12393

### CODE REVIEW

File: tools/public_api_guard/http/index.d.ts [hunk 1]
```diff
@@ -239,7 +239,7 @@ export declare class XHRConnection implements Connection {
     request: Request;
     response: Observable<Response>;
     constructor(req: Request, browserXHR: BrowserXhr, baseResponseOptions?: ResponseOptions);
-    setDetectedContentType(req: any, _xhr: any): void;
```
Comment (vicb) [inline tools/public_api_guard/http/index.d.ts]: please refrain from changing the types - add a `// TODO: ...` in the code (and open a meta issue "[3.0] Adding/fixing types" referencing the place where we should update). Thanks !

File: modules/@angular/http/src/backends/xhr_backend.ts [hunk 2]
```diff
@@ -5,15 +5,13 @@
  * Use of this source code is governed by an MIT-style license that can be
  * found in the LICENSE file at https://angular.io/license
  */
-
```
Comment (vicb) [inline modules/@angular/http/src/backends/xhr_backend.ts]: could you please leave a blank line here

File: modules/@angular/http/src/backends/xhr_backend.ts [hunk 3]
```diff
@@ -137,7 +148,7 @@ export class XHRConnection implements Connection {
     });
   }
 
-  setDetectedContentType(req: any /** TODO #9100 */, _xhr: XMLHttpRequest) {
+  setDetectedContentType(req: any /** TODO #9100 */, _xhr: any /** TODO #9100 */) {
```
Comment (vicb) [inline modules/@angular/http/src/backends/xhr_backend.ts]: why ?

File: modules/@angular/http/test/backends/xhr_backend_spec.ts [hunk 4]
```diff
@@ -502,6 +502,21 @@ export function main() {
...
+           const statusCode = 204;
+           const connection = new XHRConnection(
+               sampleRequest, new MockBrowserXHR(), new ResponseOptions({status: statusCode}));
+
+           connection.response.subscribe((res: Response) => {
+             expect(res.text()).toBe('');
+             async.done();
+           });
+
+           existingXHRs[0].setStatusCode(statusCode);
```
Comment (vicb) [inline modules/@angular/http/test/backends/xhr_backend_spec.ts]: would this have failed before ?
should you add an explicit body ?

File: modules/@angular/http/src/backends/xhr_backend.ts [hunk 5]
```diff
@@ -48,27 +49,37 @@ export class XHRConnection implements Connection {
...
-        const headers = Headers.fromResponseHeaderString(_xhr.getAllResponseHeaders());
-
-        const url = getResponseURL(_xhr);
-
         // normalize IE9 bug (http://bugs.jquery.com/ticket/1450)
         let status: number = _xhr.status === 1223 ? 204 : _xhr.status;
 
+        let body: any = null;
+
+        // 204 is not supposed to have any response body
```
Comment (vicb) [inline modules/@angular/http/src/backends/xhr_backend.ts]: nit: change to `// HTTP 204 means no content`

File: modules/@angular/http/test/backends/xhr_backend_spec.ts (hunk 4 above)
Comment (DzmitryShylovich) [inline modules/@angular/http/test/backends/xhr_backend_spec.ts]: I want to test if it will fail without fix but travis is dead :(

Comment (DzmitryShylovich) [inline modules/@angular/http/test/backends/xhr_backend_spec.ts]: done

**Pull request closed without merging**